src/api-binder.ts

```diff
@@ -0,0 +1,888 @@
+			throw e;
+		}
+	}
+
+	// Creates the necessary config vars in the API to match the current device state,
+	// without overwriting any variables that are already set.
+	private async reportInitialEnv(apiEndpoint: string) {
+		if (this.balenaApi == null) {
+			throw new InternalInconsistencyError(
+				'Attempt to report initial environment without an API client',
+			);
+		}
+
+		const targetConfigUnformatted = _.get(
+			await this.getTargetState(),
+			'local.config',
+		);
+		if (targetConfigUnformatted == null) {
+			throw new InternalInconsistencyError(
+				'Attempt to report initial state with malformed target state',
+			);
+		}
+
+		const defaultConfig = this.deviceState.deviceConfig.getDefaults();
+
+		const currentState = await this.deviceState.getCurrentForComparison();
+		const targetConfig = await this.deviceState.deviceConfig.formatConfigKeys(
+			targetConfigUnformatted,
+		);
+		const deviceId = await this.config.get('deviceId');
+
+		const currentConfig: Dictionary<string> = currentState.local.config;
+		for (const [key, value] of _.toPairs(currentConfig)) {
+			let varValue = value;
+			// We want to disable local mode when joining a cloud
+			if (key === 'SUPERVISOR_LOCAL_MODE') {
+				varValue = 'false';
+			}
+			// We never want to disable VPN if, for instance, it failed to start so far
+			if (key === 'SUPERVISOR_VPN_CONTROL') {
+				varValue = 'true';
+			}
+
+			if (targetConfig[key] == null && value !== defaultConfig[key]) {
+				const envVar = {
+					varValue,
+					device: deviceId,
+					name: 'RESIN_' + key,
+				};
+				await this.balenaApi.post({
+					resource: 'device_config_variable',
+					body: envVar,
+				});
+			}
+		}
+
+		await this.config.set({ initialConfigReported: apiEndpoint });
+	}
+
+	private async reportInitialConfig(
+		apiEndpoint: string,
+		retryDelay: number,
+	): Promise<void> {
+		try {
+			await this.reportInitialEnv(apiEndpoint);
+		} catch (err) {
+			console.error('Error reporting initial configuration, will retry', err);
+			return Bluebird.delay(retryDelay).then(() =>
+				this.reportInitialConfig(apiEndpoint, retryDelay),
+			);
+		}
+	}
+
+	private async exchangeKeyAndGetDevice(
+		opts?: KeyExchangeOpts,
+	): Promise<Device> {
+		if (opts == null) {
+			// FIXME: This casting shouldn't be necessary and stems from the
+			// meta-option provioningOptions not returning a ConfigValue
+			opts = ((await this.config.get(
+				'provisioningOptions',
+			)) as any) as KeyExchangeOpts;
+		}
+
+		const uuid = opts.uuid as string;
+		const apiTimeout = opts.apiTimeout as number;
+		if (!(uuid && apiTimeout)) {
+			throw new InternalInconsistencyError(
+				'UUID and apiTimeout should be defined in exchangeKeyAndGetDevice',
+			);
+		}
+
+		// If we have an existing device key we first check if it's
+		// valid, becaise of ot os we can just use that
+		if (opts.deviceApiKey != null) {
+			const device = await this.fetchDevice(
+				uuid,
+				opts.deviceApiKey as string,
```

Comment:
Why is this cast necessary?

Reply:
Because this is still typed as a `ConfigValue`, which is `string | number | boolean`. This is fixed with the config changes I've been working on and will be PRd as soon as this is merged.